ostree_upload_server/flatpak_importer.py

```diff
@@ -69,6 +77,116 @@ def _get_metadata_contents(repo, rev):
+        binding for this repo. The caller is expected to manage the
+        ostree transaction.
+
+        This is like "flatpak build-commit-from", but we need more
+        control over the transaction.
+        """
+        logging.info('Copying commit %s to %s', src_rev, dest_ref)
+
+        _, src_root, _ = repo.read_commit(src_rev)
+        _, src_variant, src_state = repo.load_commit(src_rev)
+
+        # Only copy normal commits
+        if src_state != 0:
+            raise Exception('Cannot copy irregular commit {}'
+                            .format(src_rev))
+
+        # If the dest ref exists, use the current commit as the new
+        # commit's parent
+        _, dest_parent = repo.resolve_rev_ext(
+            dest_ref, allow_noent=True,
+            flags=OSTree.RepoResolveRevExtFlags.REPO_RESOLVE_REV_EXT_NONE)
+        if dest_parent is not None:
+            logging.info('Using %s as new commit parent', dest_parent)
+
+        # Make a copy of the commit metadata to update. Like flatpak
+        # build-commit-from, the detached metadata is not copied since
+        # the only known usage is for GPG signatures, which would become
+        # invalid.
+        commit_metadata = GLib.VariantDict.new(src_variant.get_child_value(0))
+
+        # Set the collection binding if the repo has a collection ID,
+        # otherwise remove it
+        collection_id = FlatpakImporter._get_collection_id(repo)
+        if collection_id is not None:
+            commit_metadata.insert_value(
+                OSTree.COMMIT_META_KEY_COLLECTION_BINDING,
+                GLib.Variant('s', collection_id))
+        else:
+            commit_metadata.remove(
+                OSTree.COMMIT_META_KEY_COLLECTION_BINDING)
+
+        # Include the destination ref in the ref bindings
+        ref_bindings = commit_metadata.lookup_value(
+            OSTree.COMMIT_META_KEY_REF_BINDING,
+            GLib.VariantType('as'))
+        if ref_bindings is None:
+            ref_bindings = []
+        ref_bindings = set(ref_bindings)
+        ref_bindings.add(dest_ref)
+        commit_metadata.insert_value(
+            OSTree.COMMIT_META_KEY_REF_BINDING,
+            GLib.Variant('as', sorted(ref_bindings)))
+
+        # Add flatpak specific metadata. xa.ref is deprecated, but some
+        # flatpak clients might expect it. xa.from_commit will be used
+        # by the app verifier to make sure the commit it sent actually
+        # got there
+        commit_metadata.insert_value('xa.ref',
+                                     GLib.Variant('s', dest_ref))
+        commit_metadata.insert_value('xa.from_commit',
+                                     GLib.Variant('s', src_rev))
+
+        # Convert from GVariantDict to GVariant vardict
+        commit_metadata = commit_metadata.end()
+
+        # Copy other commit data from source commit
+        commit_subject = src_variant[3]
+        commit_body = src_variant[4]
```

Comment:
Ditto on using constant vars instead of just numbers for index offsets

Reply:
Alright.